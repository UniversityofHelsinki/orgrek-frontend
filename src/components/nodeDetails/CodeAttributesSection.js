import React from 'react';
import EditableAccordion from '../EditableAccordion';
import { useTranslation } from 'react-i18next';
import AttributesTable from '../attributes/AttributesTable';
import Validity from '../attributes/Validity';
import Placeholder from '../Placeholder';
import EditableContent from '../EditableContent';
import CodeAttributesEditor from './CodeAttributesEditor';
import {
  useGetAttributeKeysQuery,
  useSaveCodeAttributesMutation,
} from '../../store';
import { useSelector } from 'react-redux';
import { authActions } from '../../auth';
import { useNodeId } from '../../hooks/useNodeId';
import { useCodeAttributes } from '../../hooks/useCodeAttributes';
import useFilterAttributesByDate from '../../hooks/useFilterAttributesByDate';
import { compareAndCheckDates, valueNotEmpty } from './Validations';

const toFormValues = (attributes) => {
  const byKey = {};
  attributes.forEach((attribute) => {
    if (!byKey[attribute.key]) {
      byKey[attribute.key] = [];
    }
    byKey[attribute.key].push(attribute);
  });
  return byKey;
};

const withoutUniqueID = (attributes) => {
  return attributes.filter((attribute) => attribute.key !== 'unique_id');
};

const includeMissing = (attributes, allKeys) => {
  const missingKeys = allKeys.filter((key) => !attributes[key]);
  const missingAttributes = missingKeys.map((key) => ({
    id: -1,
    key: key,
    value: '',
    startDate: null,
    endDate: null,
    isNew: true,
  }));
  const missingIncluded = { ...attributes };
  missingAttributes.forEach((attribute) => {
    missingIncluded[attribute.key] = [attribute];
  });
  return missingIncluded;
};

const filterExcess = (attributes, allKeys) => {
  return attributes.filter((attribute) => {
    return allKeys.includes(attribute.key) || attribute.key === 'unique_id';
  });
};

const CodeAttributesSection = () => {
  const { t } = useTranslation();
  const nodeId = useNodeId();
  const { codeAttributes } = useCodeAttributes();
  const selectedHierarchies = useSelector(
    (s) => s.tree.selectedHierarchy || s.tree.defaultHierarchy
  );
  const presentCodeAttributes = useFilterAttributesByDate(codeAttributes);
  const { data, isFetching } = useGetAttributeKeysQuery(selectedHierarchies);
  const attributeKeys = data;

  const [saveCodeAttributes] = useSaveCodeAttributesMutation();

  if (isFetching || !attributeKeys) {
    return <></>;
  }

  const columns = [
    { label: t('code_namespace'), render: (item) => t(item.key) },
    { label: t('value'), render: (item) => item.value },
    {
      label: t('valid_dates'),
      render: (item) => (
        <Validity startDate={item.startDate} endDate={item.endDate} />
      ),
    },
  ];
  const validate = (values) => {
    const all = Object.values(values).flat();
    return {
      ...compareAndCheckDates(all),
      ...valueNotEmpty(all),
    };
  };
  const title = t('codes');
  const empty = codeAttributes.length === 0;

  const handleSubmit = (input) => {
    const attributes = Object.values(input).flat();
    const sanitized = attributes.filter((a) => !(a.isNew && a.deleted));
    return saveCodeAttributes({ attributes: sanitized, nodeId }).unwrap();
  };

  return (
    <EditableAccordion title={title}>
      <EditableContent
        editorComponent={<CodeAttributesEditor />}
        initialValues={includeMissing(
          toFormValues(
            withoutUniqueID(filterExcess(codeAttributes, attributeKeys))
          ),
          attributeKeys
        )}
        validate={validate}
        onSubmit={handleSubmit}
        successMessage={t('codeInfo.saveSuccess')}
        errorMessage={t('codeInfo.saveError')}
        authActions={authActions.codeAttributes}
      >
        <Placeholder empty={empty} placeholder={t('codeInfo.empty')}>
          <AttributesTable
            columns={columns}
            data={filterExcess(presentCodeAttributes, attributeKeys)}
            summary={title}
          />
        </Placeholder>
      </EditableContent>
    </EditableAccordion>
  );
};

export default CodeAttributesSection;
