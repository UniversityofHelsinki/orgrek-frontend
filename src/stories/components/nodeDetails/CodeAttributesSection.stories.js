import { expect } from '@storybook/jest';
import {
  mockGetAttributeKeys,
  mockGetAttributes,
  mockGetCodeAttributes,
  mockSaveCodeAttributes,
  withMockStore,
} from '../../../mockStore';
import CodeAttributesSection from '../../../components/nodeDetails/CodeAttributesSection';
import { waitFor, within } from '@storybook/testing-library';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const nodeId = '12345678';
const selectedHierarchy = 'virallinen';

export default {
  component: CodeAttributesSection,
  parameters: {
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
        hierarchies: selectedHierarchy,
      },
    },
  },
};

export const Default = {
  parameters: {
    systemTime: now,
    msw: {
      handlers: [
        mockGetAttributes(nodeId, {
          codes: [
            {
              id: 1001,
              nodeId: '1',
              key: 'lyhenne',
              value: 'LYH',
              startDate: '1970-01-01',
              endDate: null,
            },
            {
              id: 1002,
              nodeId: '1',
              key: 'emo_lyhenne',
              value: 'EMO',
              startDate: null,
              endDate: null,
            },
            {
              id: 1003,
              nodeId: '1',
              key: 'iam_ryhma',
              value: 'group1',
              startDate: null,
              endDate: null,
            },
            {
              id: 1004,
              nodeId: '1',
              key: 'talous_tunnus',
              value: '101',
              startDate: null,
              endDate: null,
            },
            {
              id: 1005,
              nodeId: '1',
              key: 'hr_lyhenne',
              value: 'HRLYH',
              startDate: null,
              endDate: null,
            },
            {
              id: 1006,
              nodeId: '1',
              key: 'hr_tunnus',
              value: 'HR102',
              startDate: null,
              endDate: null,
            },
            {
              id: 1007,
              nodeId: '1',
              key: 'tutkimus_tunnus',
              value: 'TUT102',
              startDate: null,
              endDate: null,
            },
            {
              id: 1008,
              nodeId: '1',
              key: 'oppiaine_tunnus',
              value: 'OPP102',
              startDate: null,
              endDate: null,
            },
            {
              id: 1009,
              nodeId: '1',
              key: 'laskutus_tunnus',
              value: 'LAS102',
              startDate: null,
              endDate: null,
            },
            {
              id: 1010,
              nodeId: '1',
              key: 'mainari_tunnus',
              value: 'MAI102',
              startDate: null,
              endDate: null,
            },
          ],
        }),
        mockSaveCodeAttributes(nodeId),
      ],
    },
  },
  decorators: [
    withMockStore({
      nrd: {
        node: {
          uniqueId: nodeId,
        },
      },
      tree: { selectedHierarchy },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Yksilöivä tunniste')).toBeInTheDocument();
    });

    expect(canvas.getByText('Lyhenne')).toBeInTheDocument();
    expect(canvas.getByText('Yläyksikön lyhenne')).toBeInTheDocument();
    expect(canvas.getByText('Taloustunnus')).toBeInTheDocument();
  },
};

export const Empty = {
  parameters: {
    systemTime: now,
    msw: {
      handlers: [
        mockGetAttributes(nodeId, { codes: [] }),
        mockSaveCodeAttributes(nodeId),
      ],
    },
  },
  decorators: [
    withMockStore({
      tree: { selectedHierarchy },
    }),
  ],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Yksilöivä tunniste')).toBeInTheDocument();
    });
  },
};
