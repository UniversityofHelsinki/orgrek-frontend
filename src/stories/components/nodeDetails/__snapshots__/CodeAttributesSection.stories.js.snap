// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/CodeAttributesSection Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "codes",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "edit_mode_edit_button",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "code_namespace",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "valid_dates",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "unique_id",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "12345678",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "not_specified",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "lyhenne",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "LYH",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "1.1.1970 from_date_react",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "emo_lyhenne",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "EMO",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "not_specified",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "talous_tunnus",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "101",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "not_specified",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-codeattributessection--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/CodeAttributesSection Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "codes",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "edit_mode_edit_button",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "code_namespace",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "valid_dates",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "unique_id",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "12345678",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "not_specified",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-codeattributessection--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
